packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3167 @@
+  /// [ChangeNotifier]).
+  ///
+  /// See also:
+  ///
+  /// * [of], which returns the active controller in the given context, and
+  ///   throws if one is not found.
+  static MenuController? maybeOf(BuildContext context) {
+    return context.dependOnInheritedWidgetOfExactType<_MenuHandleMarker>()?.handle._root;
+  }
+
+  // Called by MenuHandle.open to notify the controller when a menu item has
+  // been opened.
+  void _menuOpened(MenuHandle open, {required bool wasOpen}) {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    if (!wasOpen && !menuIsOpen) {
+      // We're opening the first menu, so cache the primary focus so that we can
+      // try to return to it when the menu is dismissed. Skips any focus nodes
+      // that are part of a menu system, since we don't want to return to those
+      // when the menu closes, or it will never close.
+      if (FocusManager.instance.primaryFocus?.context != null &&
+          MenuController.maybeOf(FocusManager.instance.primaryFocus!.context!) == null) {
+        assert(_debugMenuInfo('Storing previous focus as $primaryFocus'));
+        _previousFocus = FocusManager.instance.primaryFocus;
+      } else {
+        _previousFocus = null;
+      }
+    }
+    _notifyListenersSafely();
+    assert(_debugMenuInfo('Menu opened: $open'));
+  }
+
+  // Called by the _MenuNode.close to notify the controller when a menu item has
+  // been closed.
+  void _menuClosed(_MenuHandleBase close, {bool inDispose = false}) {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    if (!menuIsOpen && _previousFocus != null) {
+      // This needs to happen in the next frame so that in cases where we're
+      // closing everything, and the _previousFocus is a focus scope that
+      // currently thinks its first focus is in the menu bar, the menu bar will
+      // be unfocusable by the time the scope tries to refocus it because no
+      // menus will be open, and it will have a more appropriate first focus.
+      SchedulerBinding.instance.addPostFrameCallback((Duration _) {
+        assert(_debugMenuInfo('Returning focus to $_previousFocus'));
+        _previousFocus?.requestFocus();
+        _previousFocus = null;
+      });
+    }
+    _notifyListenersSafely();
+    assert(_debugMenuInfo('Menu closed $close'));
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<FocusNode>('previousFocus', _previousFocus));
+  }
+}
+
+/// A handle to a menu created by [createMaterialMenu].
+///
+/// A `MenuHandle` can only be created by calling [createMaterialMenu].
+///
+/// `MenuHandle` is used to control and interrogate a menu after it has been
+/// created, with methods such as [open] and [close], and state like [isOpen].
+///
+/// The [dispose] method must be called when the menu handle is no longer
+/// needed. The object shouldn't be used after [dispose] is called.
+///
```

Comment:
As commented on earlier: The distinction between MenuHandle and MenuController is not super obvious. The handle kinda feels like a controller as well. If we can't find a better solution for this, we may want to add some docs that contrast the two with each other here and in the MenuController doc.